### PR TITLE
[Intel MKL] Adding Quantized Convolution (Int8) - Part 1

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**
This PR is part 1 of four PRs that add quantized version of convolution using MKL-DNN. This part implements the quantized convolution kernel.

Please note that the below PRs replace this one #19425
Part 1: #21483 
Part 2: #21456
Part 3: #21465
Part 4: #21466

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @tatianashp, @raghuraman-k: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 21 days with no activity and the `awaiting review` label has been applied.

**Comment by mdfaijul** [comment]
@tatianashp @raghuraman-k Any update/comments on this PR?

**Review by raghuraman-k — DISMISSED** [review]

**Comment by drpngx** [comment]
@mahmoud-abuzaina could you run clang-format?

**Comment by mdfaijul** [comment]
@drpngx Ran clang-format.

**Comment by mdfaijul** [comment]
@drpngx I ran the following from my side clang-format [LLVM version 3.4.2]. Could please suggest me if I'm not doing correctly

> clang-format -i -style=Google tensorflow/core/kernels/mkl_conv_ops.cc

**Comment by mdfaijul** [comment]
kokoro:force-run

**Comment by drpngx** [comment]
Yeah unfortunately clang-format does not produce completely stable results.

**Comment by mdfaijul** [comment]
@drpngx Tests looks fine. Ready to go?

**Comment by drpngx** [comment]
I just noticed I hadn't sent the review. Sorry.

File: tensorflow/core/util/mkl_util.h [hunk 1]
```diff
@@ -104,6 +104,10 @@ typedef enum {
   Dim3d_I = 1
 } MklDnnDims3D;
 
+typedef enum {
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/util/mkl_util.h]: @drpngx None of the current PRs uses this, but upcoming PR will use this.

[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/util/mkl_util.h]: @drpngx changed enum to enum class. Thanks!

**Comment by mdfaijul** [comment]
@drpngx changed enum to enum class. Thanks!

**Comment by agramesh1** [comment]
@drpngx  Thanks for taking care of the other int8/quarantined convolution PRs. Can you approve this also?

**Comment by agramesh1** [comment]
@drpngx Sorry to bug you. Can you take a look at this PR? Thanks a lot.

**Comment by drpngx** [comment]
I'm not sure I have the expertise to review that. Maybe @tatianashp has someone in mind that would be better suited?

**Comment by agramesh1** [comment]
@drpngx  thanks. @tatianashp @raghuraman-k can you approve the changes we made since @raghuraman-k first approved the PR?

**Comment by mdfaijul** [comment]
@tatianashp @raghuraman-k @drpngx Any update on this PR?

File: tensorflow/core/util/mkl_util.h [hunk 2]
```diff
@@ -103,6 +103,10 @@ typedef enum {
   Dim3d_I = 1
 } MklDnnDims3D;
 
+enum class MklQuantization {
```
Comment (tatianashp) [inline tensorflow/core/util/mkl_util.h]: Could you please add a comment explaining what this enum means.

**Review by tatianashp — CHANGES_REQUESTED** [review]
Looks good. Just one small comment request.

File: tensorflow/core/util/mkl_util.h (hunk 2 above)
Comment (mdfaijul) [reply tensorflow/core/util/mkl_util.h]: @tatianashp Do you mean MklDnnDims3D or MklQuantization?

Comment (mdfaijul) [reply tensorflow/core/util/mkl_util.h]: @tatianashp Added comments for enums.

**Review by tatianashp — DISMISSED** [review]
Looks good. Thanks.

**Comment by mdfaijul** [comment]
@tatianashp Thanks for the approval. Th XLA errors are not related to our work, for example, similar xla failures are there for other PR (https://source.cloud.google.com/results/invocations/1756d954-ba6d-4235-b728-fe1a70ab53b4/targets)

Let us know if you need anything else for merging the PR. Thanks!

**Comment by agramesh1** [comment]
@drpngx It looks like the failures are not related to this PR. Can you start the merge process? Thanks.

**Comment by agramesh1** [comment]
@drpngx Looks like the tests are all passing now. Can you merge the PR?

**Comment by drpngx** [comment]
Merge initiated. It'll go to internal tests and approval.

**Comment by agramesh1** [comment]
@drpngx Can you take a look at this PR, it is stuck in the "ready to pull" state?

**Comment by yifeif** [comment]
The PR got imported and is pending for internal testing and approval now FYI.

**Comment by drpngx** [comment]
Could you remove the `CHECK_EQ` reformatting? There is an internal check we don't add any new `CHECK`s in the code.

**Comment by mdfaijul** [comment]
@drpngx Do you suggest any function call like `TF_RETURN_IF_ERROR`? The problem CHECK_EQ is inside some function whose return type is void and is outside OpKernel so we can't use `OP_REQUIRES` either.

**Comment by drpngx** [comment]
Either you fix it like this, which would be my preference without looking
at the code, it seems like you haven't actually changed the code, so what I
was suggesting is that you just undo the formatting change.

**Comment by mdfaijul** [comment]
@drpngx This `CHECK_EQ` is new code added in this PR. Is `TF_CHECK_OK` fine? Or  do you have any suggestion?

**Comment by drpngx** [comment]
So, you have two options:

1. You can make the function and its calling chain return `Status`.
2. If it's a true assertion, and if it's performance critical code, then you can use `DCHECK` which will only check with debug mode.

**Comment by drpngx** [comment]
Sorry, could you `DCHECK_EQ` for everything that's not an bool? `DCHECK(is_true)` is fine, but please use `DCHECK_EQ(num_value, 4)` gives a better error message.

**Review by drpngx — APPROVED** [review]
Approve new diff.

**Comment by drpngx** [comment]
Thank you!

**Comment by mahmoud-abuzaina** [comment]
@drpngx Is it passing all tests now?

**Comment by drpngx** [comment]
There is a problem with
`//tensorflow/python/autograph/pyct/static_analysis:reaching_definitions_test`. No log. Let me try again.

**Comment by mahmoud-abuzaina** [comment]
@drpngx Yes, it may be a flaky test. It is passing for me locally. Also looks like other PRs are failing because of this test too.

**Comment by drpngx** [comment]
I just kicked off the tests again, but it needs internal approval from another googler. @tatianashp could you have a look?

**Comment by agramesh1** [comment]
@drpngx thanks. @tatianashp can you approve again?

**Comment by agramesh1** [comment]
@drpngx are the tests all passing now?

**Comment by drpngx** [comment]
It fails again on the same test.

**Comment by agramesh1** [comment]
@drpngx It is not failing in our internal testing, can you provide a log? Thanks.
